src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,135 @@
+        private static readonly LocalizableString s_localizableDescription2 = new LocalizableResourceString(
+            nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadWithTwoParameterDescription),
+            MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, 
+            typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
+        internal static DiagnosticDescriptor RuleReplaceOneParameter = new DiagnosticDescriptor(
+            RuleIdOneParameterId,
+            s_localizableTitle,
+            s_localizableMessageDefault,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        internal static DiagnosticDescriptor RuleReplaceTwoParameter = new DiagnosticDescriptor(
+            RuleIdTwoParameterId,
+            s_localizableTitle2,
+            s_localizableMessageDefault2,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription2,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleReplaceOneParameter, RuleReplaceTwoParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterSyntaxNodeAction(this.Action, SyntaxKind.InvocationExpression);
+        }
+
+        private void Action(SyntaxNodeAnalysisContext context)
+        {
+            if (context.Node is InvocationExpressionSyntax invocation)
+            {
+                var invokedSymbol = context.SemanticModel.GetSymbolInfo(invocation);
+                var invokedClass = invokedSymbol.Symbol.ContainingSymbol;
+                var invokedMethod = invokedSymbol.Symbol.Name;
+                var namedType = invokedClass as INamedTypeSymbol;
+
+                if (namedType.ContainingNamespace.Name == "Text" 
+                    && namedType.ContainingNamespace.ContainingNamespace.Name == "System"
+                    && namedType.ContainingNamespace.ContainingNamespace.ContainingNamespace.IsGlobalNamespace
+                    && invokedClass.Name == "StringBuilder" 
+                    && invokedMethod == "Append"
+                    && invokedSymbol.Symbol is IMethodSymbol methodSymbol)
```

Comment:
an easier way is to use the original compilation and do `GetTypeByMetadataName("System.Text.StringBuilder")` then just use that ITypeSymbol and compare it to the one you have.

Reply:
fixed now by acquiring the necessary symbols in the StartCompilationAction.